File: examples/android/CMeSdkJava/app/src/main/java/me/connect/sdk/java/sample/proofs/ProofRequestsViewModel.java
package me.connect.sdk.java.sample.proofs;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import java.util.List;
import java.util.UUID;
import java.util.concurrent.Executors;

import me.connect.sdk.java.Connections;
import me.connect.sdk.java.Messages;
import me.connect.sdk.java.Proofs;
import me.connect.sdk.java.connection.QRConnection;
import me.connect.sdk.java.message.Message;
import me.connect.sdk.java.message.MessageState;
import me.connect.sdk.java.message.MessageType;
import me.connect.sdk.java.sample.SingleLiveData;
import me.connect.sdk.java.sample.credentials.CredentialCreateResult;
import me.connect.sdk.java.sample.db.Database;
import me.connect.sdk.java.sample.db.entity.Connection;
import me.connect.sdk.java.sample.db.entity.ProofRequest;
import me.connect.sdk.java.sample.messages.ProofDataHolder;

import static me.connect.sdk.java.sample.proofs.ProofCreateResult.FAILURE;
import static me.connect.sdk.java.sample.proofs.ProofCreateResult.FAILURE_CONNECTION;
import static me.connect.sdk.java.sample.proofs.ProofCreateResult.SUCCESS;
import static me.connect.sdk.java.sample.proofs.ProofCreateResult.SUCCESS_CONNECTION;

public class ProofRequestsViewModel extends AndroidViewModel {
    private final Database db;
    private LiveData<List<ProofRequest>> proofRequests;

    public ProofRequestsViewModel(@NonNull Application application) {
        super(application);
        db = Database.getInstance(application);
    }

    public LiveData<List<ProofRequest>> getProofRequests() {
        if (proofRequests == null) {
            proofRequests = db.proofRequestDao().getAll();
        }
        return proofRequests;
    }

    public SingleLiveData<ProofCreateResult> acceptProofRequest(int proofId) {
        SingleLiveData<ProofCreateResult> data = new SingleLiveData<>();
        acceptProofReq(proofId, data);
        return data;
    }

    private void acceptProofReqAndCreateConnection(
            ProofRequest proof,
            SingleLiveData<ProofCreateResult> liveData
    ) {
        Connections.create(proof.attachConnection, new QRConnection())
            .handle((res, throwable) -> {
                if (res != null) {
                    String serializedCon = Connections.awaitStatusChange(res, MessageState.ACCEPTED);

                    String pwDid = Connections.getPwDid(serializedCon);
                    Connection c = new Connection();
                    c.name = proof.attachConnectionName;
                    c.icon = proof.attachConnectionLogo;
                    c.pwDid = pwDid;
                    c.serialized = serializedCon;
                    db.connectionDao().insertAll(c);
                    liveData.postValue(throwable == null ? SUCCESS_CONNECTION : FAILURE_CONNECTION);

                    Proofs.retrieveAvailableCredentials(proof.serialized).handle((creds, err) -> {
                        if (err != null) {
                            liveData.postValue(FAILURE);
                            return null;
                        }
                        // We automatically map first of each provided credentials to final structure
                        // This process should be interactive in real app
                        String data = Proofs.mapCredentials(creds);
                        Proofs.send(serializedCon, proof.serialized, data, "{}").handle((s, e) -> {
                            if (s != null) {
                                String serializedProof = Proofs.awaitStatusChange(s, MessageState.ACCEPTED);
                                proof.accepted = true;
                                proof.pwDid = pwDid;
                                proof.serialized = serializedProof;
                                db.proofRequestDao().update(proof);
                            }
                            liveData.postValue(e == null ? SUCCESS: FAILURE);
                            return null;
                        });
                        return null;
                    });
                }
                if (throwable != null) {
                    throwable.printStackTrace();
                }
                return null;
            });
    }

    private void acceptProofReq(int proofId, SingleLiveData<ProofCreateResult> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            ProofRequest proof = db.proofRequestDao().getById(proofId);
            if (proof.attachConnection != null) {
                acceptProofReqAndCreateConnection(proof, liveData);
                return;
            }
            Connection con = db.connectionDao().getByPwDid(proof.pwDid);
            Proofs.retrieveAvailableCredentials(proof.serialized).handle((creds, err) -> {
                if (err != null) {
                    liveData.postValue(FAILURE);
                    return null;
                }
                // We automatically map first of each provided credentials to final structure
                // This process should be interactive in real app
                String data = Proofs.mapCredentials(creds);
                Proofs.send(con.serialized, proof.serialized, data, "{}").handle((s, e) -> {
                    if (s != null) {
                        String serializedProof = Proofs.awaitStatusChange(s, MessageState.ACCEPTED);
                        proof.accepted = true;
                        proof.serialized = serializedProof;
                        db.proofRequestDao().update(proof);
                    }
                    liveData.postValue(e == null ? SUCCESS: FAILURE);
                    return null;
                });
                return null;
            });
        });
    }

    public SingleLiveData<ProofCreateResult> rejectProofRequest(int proofId) {
        SingleLiveData<ProofCreateResult> data = new SingleLiveData<>();
        rejectProofReq(proofId, data);
        return data;
    }

    private void rejectProofReq(int proofId, SingleLiveData<ProofCreateResult> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            ProofRequest proof = db.proofRequestDao().getById(proofId);
            if (proof.pwDid == null) {
                proof.accepted = false;
                db.proofRequestDao().update(proof);
                liveData.postValue(SUCCESS);
                return;
            }
            Connection con = db.connectionDao().getByPwDid(proof.pwDid);
            Proofs.reject(con.serialized, proof.serialized).handle((s, err) -> {
                if (s != null) {
                    String serializedProof = Proofs.awaitStatusChange(s, MessageState.REJECTED);
                    proof.serialized = serializedProof;
                    proof.accepted = false;
                    db.proofRequestDao().update(proof);
                }
                liveData.postValue(err == null ? SUCCESS : FAILURE);
                return null;
            });
        });
    }

    public SingleLiveData<ProofCreateResult> getNewProofRequests() {
        SingleLiveData<ProofCreateResult> data = new SingleLiveData<>();
        checkProofRequests(data);
        return data;
    }

    private void checkProofRequests(SingleLiveData<ProofCreateResult> data) {
        Executors.newSingleThreadExecutor().execute(() -> {
            Messages.getPendingMessages(MessageType.PROOF_REQUEST, null, null).handle((res, throwable) -> {
                if (throwable != null) {
                    throwable.printStackTrace();
                }
                if (res != null) {
                    for (Message message : res) {
                        ProofDataHolder holder = ProofDataHolder.extractRequestedFieldsFromProofMessage(message);
                        String pwDid = message.getPwDid();
                        if (!db.proofRequestDao().checkExists(holder.threadId)) {
                            Proofs.createWithRequest(UUID.randomUUID().toString(), holder.proofReq).handle((pr, err) -> {
                                if (err != null) {
                                    err.printStackTrace();
                                } else {
                                    ProofRequest proof = new ProofRequest();
                                    proof.serialized = pr;
                                    proof.name = holder.name;
                                    proof.pwDid = pwDid;
                                    proof.attributes = holder.attributes;
                                    proof.threadId = holder.threadId;
                                    proof.messageId = message.getUid();
                                    db.proofRequestDao().insertAll(proof);

                                    Messages.updateMessageStatus(pwDid, message.getUid());
                                }
                                return null;
                            });
                        }
                    }
                }
                data.postValue(SUCCESS);
                return null;
            });
        });
    }
}

File: examples/android/CMeSdkJava/app_kt/src/main/kotlin/me/connect/sdk/java/samplekt/proofs/ProofRequestsViewModel.kt
package me.connect.sdk.java.samplekt.proofs

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.future.await
import kotlinx.coroutines.launch
import me.connect.sdk.java.Connections
import me.connect.sdk.java.Messages
import me.connect.sdk.java.Proofs
import me.connect.sdk.java.connection.QRConnection
import me.connect.sdk.java.message.MessageState
import me.connect.sdk.java.message.MessageType
import me.connect.sdk.java.samplekt.SingleLiveData
import me.connect.sdk.java.samplekt.db.Database
import me.connect.sdk.java.samplekt.db.entity.Connection
import me.connect.sdk.java.samplekt.db.entity.ProofRequest
import me.connect.sdk.java.samplekt.messages.ProofDataHolder
import me.connect.sdk.java.samplekt.proofs.ProofCreateResult.*
import me.connect.sdk.java.samplekt.wrap
import java.util.*


class ProofRequestsViewModel(application: Application) : AndroidViewModel(application) {
    private val db: Database = Database.getInstance(application)
    private val proofRequestsLiveData by lazy {
        db.proofRequestDao().getAll()
    }

    fun getProofRequests(): LiveData<List<ProofRequest>> = proofRequestsLiveData

    fun acceptProofRequest(proofId: Int): SingleLiveData<ProofCreateResult> {
        val data = SingleLiveData<ProofCreateResult>()
        acceptProofReq(proofId, data)
        return data
    }

    private suspend fun acceptProofReqAndCreateConnection(
        proof: ProofRequest,
        liveData: SingleLiveData<ProofCreateResult>
    ) {
        val res = Connections.create(proof.attachConnection!!, QRConnection()).wrap().await()
        if (res != null) {
            val serializedCon = Connections.awaitStatusChange(res, MessageState.ACCEPTED)
            val pwDid = Connections.getPwDid(serializedCon)
            val c = Connection(
                name = proof.attachConnectionName!!,
                icon = proof.attachConnectionLogo,
                pwDid = pwDid,
                serialized = serializedCon
            )
            proof.pwDid = pwDid
            db.connectionDao().insertAll(c)
            liveData.postValue(SUCCESS_CONNECTION)
            db.proofRequestDao().update(proof)
            var creds: String? = null
            try {
                creds = Proofs.retrieveAvailableCredentials(proof.serialized).wrap().await()
            } catch (e: Exception) {
                liveData.postValue(MISSED)
            }
            val data = Proofs.mapCredentials(creds)
            val s = Proofs.send(serializedCon, proof.serialized, data, "{}").wrap().await()
            if (s != null) {
                val serializedProof =
                    Proofs.awaitStatusChange(s, MessageState.ACCEPTED)
                proof.accepted = true
                proof.serialized = serializedProof
                db.proofRequestDao().update(proof)
                liveData.postValue(SUCCESS)

            } else {
                liveData.postValue(FAILURE)
            }
        } else {
            liveData.postValue(FAILURE_CONNECTION)
        }
    }

    private fun acceptProofReq(proofId: Int, liveData: SingleLiveData<ProofCreateResult>) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val proof: ProofRequest = db.proofRequestDao().getById(proofId)
            if (proof.attachConnection != null && proof.pwDid == null) {
                acceptProofReqAndCreateConnection(proof, liveData)
                return@launch
            }
            val con: Connection = db.connectionDao().getByPwDid(proof.pwDid!!)
            val creds: String? = null
            try {
                val creds = Proofs.retrieveAvailableCredentials(proof.serialized).wrap().await()
            } catch (e: Exception) {
                liveData.postValue(MISSED)
            }
            val data = Proofs.mapCredentials(creds)
            val s =  Proofs.send(con.serialized, proof.serialized, data, "{}").wrap().await()
            if (s != null) {
                val serializedProof =
                    Proofs.awaitStatusChange(s, MessageState.ACCEPTED)
                proof.accepted = true
                proof.serialized = serializedProof
                db.proofRequestDao().update(proof)
            }
            liveData.postValue(SUCCESS)
        } catch (e: Exception) {
            e.printStackTrace()
            liveData.postValue(FAILURE)
        }

    }

    fun rejectProofRequest(proofId: Int): SingleLiveData<ProofCreateResult> {
        val data = SingleLiveData<ProofCreateResult>()
        rejectProofReq(proofId, data)
        return data
    }

    private fun rejectProofReq(proofId: Int, liveData: SingleLiveData<ProofCreateResult>) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val proof = db.proofRequestDao().getById(proofId)
            val con = db.connectionDao().getByPwDid(proof.pwDid!!)
            val s = Proofs.reject(con.serialized, proof.serialized).wrap().await()
            val serializedProof = Proofs.awaitStatusChange(s, MessageState.REJECTED)
            proof.serialized = serializedProof
            proof.accepted = false
            db.proofRequestDao().update(proof)
            liveData.postValue(SUCCESS)
        } catch (e: Exception) {
            e.printStackTrace()
            liveData.postValue(FAILURE)
        }
    }

    fun getNewProofRequests(): SingleLiveData<ProofCreateResult> {
        val data = SingleLiveData<ProofCreateResult>()
        checkProofRequests(data)
        return data
    }

    private fun checkProofRequests(data: SingleLiveData<ProofCreateResult>) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val res = Messages.getPendingMessages(MessageType.PROOF_REQUEST, null, null).wrap().await()

            res.forEach { message ->
                val holder = ProofDataHolder.extractRequestedFieldsFromProofMessage(message)!!
                val pwDid: String = message.pwDid
                if (!db.proofRequestDao().checkExists(holder.threadId)) {
                    val pr = Proofs.createWithRequest(UUID.randomUUID().toString(), holder.proofReq).wrap().await()
                    val proof = ProofRequest(
                            serialized = pr,
                            name = holder.name,
                            pwDid = pwDid,
                            attributes = holder.attributes,
                            threadId = holder.threadId,
                            messageId = message.uid
                    )
                    db.proofRequestDao().insertAll(proof)

                    Messages.updateMessageStatus(pwDid, message.uid)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            data.postValue(SUCCESS)
        }
    }
}
